File: backend/app/models/flow.py
"""Flow 自动化模型（M06；dev-plan P14；api-spec §3.7）。"""

from __future__ import annotations

from datetime import datetime

from sqlalchemy import DateTime, Index, Integer, Text
from sqlalchemy.orm import Mapped, mapped_column

from app.models.base import Base

TRIGGER_TYPES = ("cron", "webhook", "manual", "event")


class Flow(Base):
    __tablename__ = "flows"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    name: Mapped[str] = mapped_column(Text)
    description: Mapped[str] = mapped_column(Text, default="")
    trigger_type: Mapped[str] = mapped_column(Text, default="manual")
    trigger_config: Mapped[str] = mapped_column(Text, default="{}")  # JSON
    actions: Mapped[str] = mapped_column(Text, default="[]")  # JSON 动作数组（含条件节点）
    enabled: Mapped[int] = mapped_column(Integer, default=0)
    webhook_token: Mapped[str | None] = mapped_column(Text, default=None, unique=True)
    retry: Mapped[int] = mapped_column(Integer, default=0)
    retry_interval: Mapped[int] = mapped_column(Integer, default=60)
    last_run_at: Mapped[datetime | None] = mapped_column(DateTime, default=None)


class FlowRun(Base):
    __tablename__ = "flow_runs"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    flow_id: Mapped[int] = mapped_column(Integer)
    trigger: Mapped[str] = mapped_column(Text)  # cron/webhook/manual/event
    status: Mapped[str] = mapped_column(Text, default="running")  # running/success/failed
    steps_log: Mapped[str] = mapped_column(Text, default="[]")  # JSON 每步输入输出
    started_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)
    finished_at: Mapped[datetime | None] = mapped_column(DateTime, default=None)
    duration_ms: Mapped[int | None] = mapped_column(Integer, default=None)

    __table_args__ = (Index("ix_flow_runs_flow_ts", "flow_id", "started_at"),)

File: frontend/src/api/flow.ts
import request from './request'
import type { FlowGraph } from '../utils/canvas'

/** Flow 自动化·表单版 + 画布（M06；dev-plan P14/P19；api-spec §4.7）。 */

export type FlowTrigger = 'cron' | 'webhook' | 'manual' | 'event'
export type FlowActionType = 'http' | 'notify' | 'condition' | 'ssh' | 'docker' | 'ai' | 'delay' | 'variable'

export interface FlowAction {
  type: FlowActionType
  name?: string
  expression?: string
  config?: Record<string, unknown>
}

export interface FlowItem {
  id: number
  name: string
  description: string
  trigger_type: FlowTrigger
  trigger_config: Record<string, unknown>
  actions: FlowAction[]
  graph: FlowGraph | null
  enabled: boolean
  webhook_token: string | null
  retry: number
  retry_interval: number
  last_run_at: string | null
}

export interface FlowRunItem {
  id: number
  flow_id: number
  trigger: string
  status: 'running' | 'success' | 'failed'
  steps: Array<Record<string, unknown>>
  started_at: string | null
  finished_at: string | null
  duration_ms: number | null
}

export interface FlowBody {
  name: string
  description: string
  trigger_type: FlowTrigger
  trigger_config: Record<string, unknown>
  actions: FlowAction[]
  graph?: FlowGraph | null
  enabled: boolean
  retry: number
  retry_interval: number
}

export interface FlowTemplate {
  key: string
  name: string
  description: string
  trigger_type: FlowTrigger
  trigger_config: Record<string, unknown>
  has_canvas: boolean
}

export const flowApi = {
  list: () => request.get<never, FlowItem[]>('/flows'),
  get: (id: number) => request.get<never, FlowItem>(`/flows/${id}`),
  create: (body: FlowBody) => request.post<never, FlowItem>('/flows', body),
  update: (id: number, body: FlowBody) => request.put<never, FlowItem>(`/flows/${id}`, body),
  remove: (id: number) => request.delete(`/flows/${id}`),
  run: (id: number) => request.post<never, { run_id: number; status: string }>(`/flows/${id}/run`),
  dryRun: (id: number) =>
    request.post<never, { run_id: number; status: string; steps: Array<Record<string, unknown>> }>(
      `/flows/${id}/dry-run`,
    ),
  runs: (id: number, limit = 20) => request.get<never, FlowRunItem[]>(`/flows/${id}/runs`, { params: { limit } }),
  runDetail: (runId: number) => request.get<never, FlowRunItem>(`/flow-runs/${runId}`),
  templates: () => request.get<never, FlowTemplate[]>('/flows/templates'),
  fromTemplate: (key: string, name?: string) =>
    request.post<never, FlowItem>('/flows/from-template', { key, name }),
  exportFlow: (id: number) =>
    request.get<never, Record<string, unknown>>(`/flows/${id}/export`),
  importFlow: (payload: Record<string, unknown>) =>
    request.post<never, FlowItem>('/flows/import', payload),
  resetToken: (id: number) => request.post<never, { webhook_token: string }>(`/flows/${id}/reset-token`),
}
